### DISCUSSION TITLE
Is it possible to run sionna rt with multiple GPUs？

### DISCUSSION THREAD

**Discussion by yzw0905** [NVlabs/sionna#231] [Q&A] [answered]
Hi, in the docs I see that sionna rt is limited to running with a single GPU, I wonder if it is possible to run with multiple GPUs at the same time to improve efficiency? Multiple GPUs seem to be related to training, but these are not seen in the sionna rt code.

**Comment by jhoydis** [comment]
Hi,

It is currently not possible to use multiple GPUs to speed-up ray tracing, e.g., for a very large number of receivers/transmitters.
However, you can load the same scene on different GPUs and do the ray tracing for different sets of receivers/transmitters in parallel.

**Reply by yzw0905** [reply]
thank you for your reply！I also have a question, set multiple tx in covermap, will there be an effect between them, because I noticed that the num_samples described in the code was split equally by tx, so I ran a single transmitter num_samples 10e6 and two transmitters num_samples 20e6 covermap, for the initial launch of the resulting image is different, this question arises ...

**Reply by jhoydis** [reply]
Hi,

The number of samples (i.e., rays) is equally divided between the transmitters in the scene. There should hence be no difference in the coverage maps for 1 TX and 10e6 samples and 2 TX and 20e6 samples.

However, if you enable scattering, there can be slight differences as the output of the ray tracer is not deterministic.

**Reply by yzw0905** [reply]
Thanks for your reply, I thought the same way, but below are the images corresponding to 1tx 10e6 and 2tx 20e6, they are not quite consistent, I only turned on the default reflection.
![单10e6](https://github.com/NVlabs/sionna/assets/132966337/c3b6f696-6217-4ff1-9036-aabbd49c957d)
![双20e6](https://github.com/NVlabs/sionna/assets/132966337/d37568c3-27c6-4c27-a899-eef3511f14ee)
Here are my covermap parameter settings：
![image](https://github.com/NVlabs/sionna/assets/132966337/deb34e44-631a-453e-9bdb-981bb4657d93)

**Reply by jhoydis** [reply]
Hi,

Could you share a full code example that would allow us to reproduce the issue?

**Reply by yzw0905** [reply]
Thank you really for your answer！ here is all the code I used
```
import os
gpu_num = 0 # Use "" to use the CPU
os.environ["CUDA_VISIBLE_DEVICES"] = f"{gpu_num}"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

# Colab does currently not support the latest version of ipython.
# Thus, the preview does not work in Colab. However, whenever possible we
# strongly recommend to use the scene preview mode.
try: # detect if the notebook runs in Colab
    import google.colab
    colab_compat = True # deactivate preview
except:
    colab_compat = False
resolution = [480,320] # increase for higher quality of renderings

# Allows to exit cell execution in Jupyter
class ExitCell(Exception):
    def _render_traceback_(self):
        pass

# Import Sionna
try:
    import sionna
except ImportError as e:
    # Install Sionna if package is not already installed
    import os
    os.system("pip install sionna")
    import sionna

# Configure the notebook to use only a single GPU and allocate only as much memory as needed
# For more details, see https://www.tensorflow.org/guide/gpu
import tensorflow as tf
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    try:
        tf.config.experimental.set_memory_growth(gpus[0], True)
    except RuntimeError as e:
        print(e)
# Avoid warnings from TensorFlow
tf.get_logger().setLevel('ERROR')

tf.random.set_seed(1) # Set global random seed for reproducibility

import matplotlib.pyplot as plt
import numpy as np
import time
# Import Sionna RT components
from sionna.rt import load_scene, Transmitter, Receiver, PlanarArray, Camera
```
```
# Load integrated scene
scene = load_scene(sionna.rt.scene.bybs) # Try also sionna.rt.scene.etoile
# Configure antenna array for all transmitters
scene.tx_array = PlanarArray(num_rows=1,
                             num_cols=1,
                             vertical_spacing=0.5,
                             horizontal_spacing=0.5,
                             pattern="tr38901",
                             polarization="V")

# Configure antenna array for all receivers
scene.rx_array = PlanarArray(num_rows=1,
                             num_cols=1,
                             vertical_spacing=0.5,
                             horizontal_spacing=0.5,
                             pattern="dipole",
                             polarization="cross")

# Create transmitter
tx = Transmitter(name="tx",
                 position=[100,100,70])

# Add transmitter instance to scene
scene.add(tx)

# Create a receiver
rx = Receiver(name="rx",
              position=[-300,-30,20],
              orientation=[0,0,0])

# Add receiver instance to scene
scene.add(rx)

tx.look_at(rx) # Transmitter points towards receiver
```
```
tx = Transmitter(name="tx1",
                 position=[400,40,80], 
                 orientation=[0,0,0])
scene.add(tx)


tx = Transmitter(name="tx2",
                 position=[-200,-67,40], 
                 orientation=[0,0,0])
scene.add(tx)
```
```
cm = scene.coverage_map(max_depth=5,
                        cm_cell_size=(1,1), # Grid size of coverage map cells in m
                        combining_vec=None,
                        precoding_vec=None,
                        num_samples=int(20e6)) # Reduce if your hardware does not have enough memory
```

**Reply by jhoydis** [reply]
Hi,

There seems to be indeed an issue. We look into it. Thanks for reporting this.

**Reply by yzw0905** [reply]
Thanks！ that's what I should have done

**Comment by faycalaa** [comment, marked as answer]
Hi,

There will be a fix in the next release.

In case you don't want to wait, the fix consists in a single line of code: You need to add

```
samples_tx_indices = tf.boolean_mask(samples_tx_indices, active)
```

after [line 1849 in sionna/rt/solver_cm.py](https://github.com/NVlabs/sionna/blob/f5ea373b948bfc367eaccdb8907889cfe62badd9/sionna/rt/solver_cm.py#L1849).

**Reply by yzw0905** [reply]
HI,
Thanks, I'll revise it based on your suggestion

**Comment by faycalaa** [comment]
Fixed by v0.16.
